fix autodiff macro on generic functions

haenoe opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

heloo there!
This short PR allows applying the `autodiff` macro to generic functions like this one. 
It only touches the frontend part, since the `rustc_autodiff` macro can already handle generics.
```rust
#[autodiff(d_square, Reverse, Duplicated, Active)]
fn square<T: std::ops::Mul<Output = T> + Copy>(x: &T) -> T {
    *x * *x
}
```
Thanks to Manuel for creating an issue on this. For more information on this see #140032 
r? @ZuseZ4 

As always: thanks for any piece of feedback!!

Fixes: #140032
Tracking issue for autodiff: https://github.com/rust-lang/rust/issues/124509

haenoe committed:
feat: propagate generics to generated function

haenoe committed:
feat: add test for generics in generated function

haenoe commented:
> ...
> The source function square is fine, but d_square has to change. We need to copy the generic bounds. We might also want to specify the bounds in the body (the `::<T>`), but I'm not sure if that's ever needed, so we could also skip it in the beginning. 
> ```rs
> #[rustc_autodiff(Reverse, 1, Duplicated, Active)]
> #[inline(never)]
> fn d_square<T: std::ops::Mul<Output = T> + Copy>(x: &T, dx_0: &mut T, dret: T) -> T {
>     unsafe {
>         asm!("NOP", options(pure, nomem));
>     };
>     ::core::hint::black_box::<T>(square(x));
>     ::core::hint::black_box((dx_0, dret));
>     ::core::hint::black_box::<T>(square(x))
> }
> ``` 

 _Originally posted by @ZuseZ4 in [#140032](https://github.com/rust-lang/rust/issues/140032#issuecomment-2816617933)_

As mentioned by Manuel here, it may also be interesting to specify the type bounds in the generated body.
Should we implement this as a preliminary step? Happy about any input on this!

ZuseZ4 commented:
That were few lines of code, nice. Can you also add a test to `tests/codegen/autodiff` to make sure that we keep handling monomorphization correctly? Especially, please add a test where you call d_square with a type which you haven't used when calling square, to make sure that just calling d_square will already force the correct instantiation of square.


Also, can you please add the turbofish when calling the functions in the body? 
@kpreid just gave a good example of when it would be needed. If foo is defined with some generic that isn't used on neither inputs, nor the return type, then rustc won't infer it if we call foo. So it would be good to specify it and add a test for that.

haenoe committed:
feat: add generated parameters to generated function

- update pretty printing tests
- only add generic parameters when function is actually generic (no empty turbofish)

haenoe commented on compiler/rustc_builtin_macros/src/autodiff.rs:
```diff
@@ -674,8 +678,10 @@ mod llvm_enzyme {
         span: Span,
         primal: Ident,
         idents: &[Ident],
+        generics: &Generics,
```
Passing this through 3x is a bit ugly :/

ZuseZ4 replied:
There is a discussion about fully dropping the function body and just generating a declaration using `#[rustc_intrinsic]` ^^ [#wg-autodiff > Placeholder function design @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/390790-wg-autodiff/topic/Placeholder.20function.20design/near/510446645)
This would make the frontend trivial, but complicate things a bit in the middle or backend. I think it's still an overall win, but I haven't looked into it.

haenoe commented on compiler/rustc_builtin_macros/src/autodiff.rs:
```diff
@@ -674,8 +678,10 @@ mod llvm_enzyme {
     ) -> P<ast::Expr> {
         let has_self = idents.len() > 0 && idents[0].name == kw::SelfLower;
+
         if has_self {
```
Right now, I left the code only in the lower branch.
@ZuseZ4 could you quickly explain why we have this explicit branching here?

ZuseZ4 replied:
I assume it's for methods, e.g. https://github.com/rust-lang/rust/issues/139557
But we don't have tests in `tests/pretty/` using self, and as per issue I also broke support for it at some point, so .. that checks out. 

I think that this code handles calling another method in the dummy body, so instead of generating
`bench_black_box(d_psi(self, j, 1.0));` we'd generate `bench_black_box(self.d_psi(j, 1.0));`

haenoe replied:
Since the `impl` PR is already being rolled up I will try to fix it as part of this PR after I have rebased

haenoe commented on compiler/rustc_builtin_macros/src/autodiff.rs:
```diff
@@ -684,7 +690,46 @@ mod llvm_enzyme {
+                    .iter()
+                    .filter(|param| matches!(param.kind, ast::GenericParamKind::Type { .. }));
+
+                let generated_generic_types = primal_generic_types
```
There probably is some helper functions available in `ExtCtxt` make this less ugly? ^^'

haenoe commented on tests/pretty/autodiff_forward.pp:
```diff
@@ -39,7 +39,7 @@
 pub fn df1(x: &[f64], bx_0: &[f64], y: f64) -> (f64, f64) {
     unsafe { asm!("NOP", options(pure, nomem)); };
-    ::core::hint::black_box(f1(x, y));
+    ::core::hint::black_box(f1::<>(x, y));
```
Do we want the `::<>` in non generic functions, or just emit it fully?

ZuseZ4 replied:
I'd probably skip it in non-generic functions. I don't think too many people will look at it, but I still try to keep the function somewhat simple.

haenoe commented:
Left some of the things I though about while writing -- will head to bed now! Thanks for any tips!

bors commented:
:umbrella: The latest upstream changes (presumably #140165) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

haenoe commented:
> That were few lines of code, nice. Can you also add a test to `tests/codegen/autodiff` to make sure that we keep handling monomorphization correctly? Especially, please add a test where you call d_square with a type which you haven't used when calling square, to make sure that just calling d_square will already force the correct instantiation of square.

@ZuseZ4 do I understand correctly, that when I call `d_square::<f64>` without having called `square::<f64>` before, then `square::<f64>` should be generated? Or should the last "square" in your statement be in fact "d_square"?

haenoe committed:
feat: add codegen test

Ensure that code for generic `d_primal::<T>` is generated even if `primal::<T>`
is never used.

- incorporate feedback from @ZuseZ4

haenoe commented:
Sorry, wrong rebase xD

ZuseZ4 commented:
Yes exactly, the macro should internally generate the primal call and force instantiation. So that should be tested by not manually calling the primal.

haenoe commented:
Okay! I can now better explain why I was confused. I copied the test parameters from one of the other codegen tests (which included `opt-level=3`.
[This gist](https://gist.github.com/haenoe/f426109f6c8e31985284dd8127acd459) shows the LLVM-IR output of the test I added
1. with `opt-level=0`
2. with `opt-level=3`

When searching for `generic::square` the `opt-level=0` version, you can see that indeed, the macro has caused the compiler to generate two versions of the function (one which takes `float`, and another which takes `double`).
In the `opt-level=3` output however, the function is only generated once for `float`.

Is the compiler smart enough to find out that `square::<f64>` is _actually_ not needed or could this pose a problem for us?
Thanks for you help!

ZuseZ4 commented:
Due to our macro generating a call to the primal function, ristc will instantiate it. Then we run Enzyme and differentiate the primal, and replace the placeholder body with the actual derivative code. That means that the call to our primal, which enzyme generated, is gone. We run another opt pipeline after AD, so llvm will recognize the now unused primal and delete it under O3.
If you run optlevel=3 with NoPostopt flag (might be in my PR which is currently in the merge queue), them you should still see the primal in the codegen test (I didn't try it, so correct me if I'm wrong)

On April 23, 2025 12:15:23 PM EDT, max ***@***.***> wrote:
>haenoe left a comment (rust-lang/rust#140049)
>
>Okay! I can now better explain why I was confused. I copied the test parameters from one of the other codegen tests (which included `opt-level=3`.
>[This gist](https://gist.github.com/haenoe/f426109f6c8e31985284dd8127acd459) shows the LLVM-IR output of the test I added
>1. with `opt-level=0`
>2. with `opt-level=3`
>
>When searching for `generic::square` the `opt-level=0` version, you can see that indeed, the macro has caused the compiler to generate two versions of the function (one which takes `float`, and another which takes `double`).
>In the `opt-level=3` output however, the function is only generated once for `float`.
>
>Is the compiler smart enough to find out that `square::<f64>` is _actually_ not needed or could this pose a problem for us?
>Thanks for you help!
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/rust-lang/rust/pull/140049#issuecomment-2824833415
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

ZuseZ4 commented on tests/codegen/autodiff/generic.rs:
```diff
@@ -0,0 +1,37 @@
+    //              f32
+    //             | | |
+    //             V V V
+    // CHECK: fmul float %{{.+}}, %{{.+}}
```
This is too generic, as it can match an arbitrary float multiplication anywhere in the rust std lib (since we use fat-lto). You can check some of the other codegen tests for example.
A more reliable setup would be
1) CHECK the debug name of the function 
2) CHECK-NEXT: the beginning of the function header (ignore anything after the closing bracket after the last arg)
3) CHECK-NOT: ret
4) CHECK: fmul float ...
Step 3 is to make sure that we haven't seen a return statement yet, and thus that the fmul is still part of the same function.

(and do that for both cases). I'd also move the CHECK's out of the main function just so that they are at the outermost level. I'm not sure if it's required by other tooling, but that's the convention I've usually seen (correct me if you saw some other examples).

ZuseZ4 replied:
ping @haenoe, do you have some time to look into it, or do you prefer it if I write the tests?

haenoe replied:
Heyy Manuel!
Sorry, had a lot going on. I will look at it tomorrow at the latest.
One thing that felt kind of weird was matching the debug name, which are random characters the name and then some more random chars.

Do you think a `.*name.*` is sufficient?

ZuseZ4 replied:
No worries, I can always take over, even partial PRs are helpful. I just want to make sure that this one doesn't end up stalled because people seem to be excited to differentiate generic functions.

Yes it's sufficient because we print the debug name as a comment above the function. In LLVM-IR, you can start a comment with `;`, it's not the `//` from C++ or Rust. So in our LLVM-IR we have

```llvm
; moduleName::functionName
define mangledFnName(...)
```
So you wouldn't match the mangled name, but the real name above. With `CHECK-NEXT` you then make sure that you match the real function on the following line.

haenoe commented:
Hey thanks a lot for your patience and esp. your assistance :D

ZuseZ4 commented:
(I'm currently at rustweek, but also want to look at the full generated llvm-ir before merging, so I might take a few days longer to review this time)

ZuseZ4 commented:
Thanks! Yes it was nice, hopefully you get a chance to join next year.
Ok, during my flight home I finally got to build it locally and check the IR, lgtm.

@bors r+ rollup

bors commented:
:pushpin: Commit a504759afd3aa8338ebcb4ddc7b042de14a29642 has been approved by `ZuseZ4`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ZuseZ4 a504759afd3aa8338ebcb4ddc7b042de14a29642 -->
<!-- homu: {"type":"Approved","sha":"a504759afd3aa8338ebcb4ddc7b042de14a29642","approver":"ZuseZ4","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
